### ISSUE TITLE
Roadmap: EDOPro adapter, privacy, decision projection, model integration, and release

### ISSUE THREAD

**Issue by chrismaghuhn** [chrismaghuhn/OCGForge-Ignis#3]
# Goal

Track OCGForge-Ignis from the accepted clean-room protocol foundation through a deterministic local/private EDOPro adapter, perspective-safe state, complete decision projection, byte-exact OCGForge compatibility, model-runner integration, local duel coverage, audit tooling, and Windows release readiness.

Related OCGForge core / ML roadmap:

- https://github.com/chrismaghuhn/OCGForge/issues/16

This issue is a **roadmap/reminder tracker**, not task authorization. Every implementation slice still follows:

```text
ONE TASK
→ implement
→ focused/local gates
→ commit + push
→ STOP
→ independent live-repository review
→ fix if needed
→ FINAL PASS
→ only then authorize the next task
```

Priority:

```text
correctness
→ determinism
→ information safety
→ complete legal decisions
→ replay / auditability
→ maintainability
→ performance
→ ML scale
```

---

# Current authoritative status — 2026-09-03

Accepted `main`:

```text
fff3269918f7f9120e815b900d8cc2e14e1bc52d
```

Accepted:

```text
I0 FINAL PASS
I1 FINAL PASS
```

Current I2 PR:

- https://github.com/chrismaghuhn/OCGForge-Ignis/pull/2
- reviewed head: `087c46db4fab638ed4c0055ac2a49fcc621b8cca`
- PR open / not merged
- independent review found remediation requirements

```text
I2_FINAL_PASS=NO
MERGE_AUTHORIZED=NO
I3_AUTHORIZED=NO
```

The current I2 remediation scope covers TCP-chunk-independent semantic barriers, causal packet boundaries around pre-duel choices, transport-close explicit leave semantics, and recovery from authoritative lobby invalidation while a host start request is pending.

---

# Frozen architecture

OCGForge-Ignis is a separate standalone external EDOPro network client.

```text
local/private EDOPro
        ↕ TCP / CTOS-STOC
OCGForge-Ignis.exe
        ├─ protocol
        ├─ client/lobby
        ├─ perspective state/privacy
        ├─ decision projection
        ├─ continuation handling
        └─ audit trace
        ↕ framed child-process IPC
OCGForge model runner
        ↕
exact pinned canonical checkpoint
```

Frozen boundaries:

```text
LANGUAGE=C# / .NET 10 LTS
EDOPRO_FORK_REQUIRED=NO
EDOPRO_PLUGIN_REQUIRED=NO
WINDBOT_CODE_REUSE=NO
OCGCORE_EMBEDDING=NO
DLL_INJECTION=FORBIDDEN
SEPARATE_EXE=YES
SEPARATE_MODEL_RUNNER=YES
```

EDOPro remains the only gameplay/legal authority for the external duel.

The adapter may translate a proven current prompt, but may not:

- recompute general Yu-Gi-Oh! legality;
- add/remove/truncate/sort/deduplicate/repair legal options;
- invent Pass/Cancel;
- auto-answer N=1;
- fall back to Teacher/RandomLegal/first candidate;
- retry failed model decisions under another policy.

---

# V1 target

Initial scope:

```text
EDOPro source pin = 30935e847165a9ef0e547fb51a43f36168fab7c7
EDOPro version target = 41.0.2 "Bagooska"
PRO_VERSION = 0x1354
EDOPro ocgcore gitlink = 46779fbe40e6a9bd8967f5dc6a03f4eaa6550d57
ocgcore API = 11
local/private pre-created rooms only
1v1 only
single duel only
```

Explicit V1 exclusions include public production servers, Match/Siding, Tag/Relay, observer operation, reconnect/catchup, rematch, room creation/discovery, EDOPro forks/plugins/hooks, arbitrary ANNOUNCE_CARD reconstruction, and training from raw protocol/human match logs.

Public Project-Ignis server use remains:

```text
OFFICIAL_PUBLIC_SERVER_USE=NOT_AUTHORIZED
OI-N01=CLOSED
```

Endpoint availability is not bot permission.

---

# Privacy architecture

Keep distinct layers:

```text
RawProtocolState
PerspectiveStateMirror
PublicContractProjection
PrivateResponseBinding
ModelRunnerBoundary
PublicAuditTrace
PrivateProtocolTrace
```

Opponent hidden card identities, raw packet offsets, socket/session state, room passwords, private response bindings, pointers/object IDs, PID, wall time, absolute paths, and other control-plane facts must never enter authoritative model input or semantic gameplay identity.

Knowledge-destroying transitions such as shuffle must destroy stale hidden locators.

Private protocol traces remain:

```text
TRAINING_ELIGIBILITY=NO
AUTHORITATIVE_MODEL_INPUT=NO
```

Paired-hidden-world tests must preserve byte equality through public projection and later model scoring/selection whenever only hidden opponent facts differ.

---

# Implementation stages

## I0 — governance / threat model / upstream pins

**FINAL PASS**

Repository charter, architecture, privacy threat model, runtime identity, clean-room provenance, exact upstream roles/pins, UI/headless ownership boundary.

Accepted I0 head:

```text
4a7819b5fbe1d0c989954d9d035bbf561a0833fe
```

## I1 — deterministic CTOS/STOC wire codec

**FINAL PASS / merged**

Owns framing, LE primitives, typed packet DTOs, exact fixed UTF-16 semantics, padding/non-semantic bytes, discriminated errors, version compatibility validation, fragmentation/coalescing determinism.

Accepted I1 head:

```text
914ec14f8e9269a877042563ebaa956522e0bc9c
```

Merged main:

```text
fff3269918f7f9120e815b900d8cc2e14e1bc52d
```

## I2 — connection / lobby / pre-duel

**CURRENT / remediation required**

Owns:

- production `TcpClient` + `NetworkStream` transport;
- deterministic scripted transport tests;
- explicit local/private endpoint configuration;
- player-info/join/deck/ready/start pre-duel control;
- 1v1/single-duel topology gates;
- server-authoritative ready state;
- RPS / turn-preference handshake;
- exact transport + unread-suffix handoff before gameplay parsing;
- no retry/reconnect/fallback.

I2 must hand off with:

```text
FINAL_GAMEPLAY_PERSPECTIVE=UNRESOLVED_AT_I2_HANDOFF
```

The final perspective is established later from the first `MSG_START`, not from lobby position.

## I3 — perspective state mirror + privacy projection

Blocked until I2 FINAL PASS + merge.

Owns perspective-safe public state reconstruction from the bot's received protocol stream:

- public zones/counts;
- visible card identities/properties;
- known/unknown state;
- chain/public relationships;
- deterministic semantic locators;
- knowledge-destroying transitions;
- first `MSG_START` gameplay perspective establishment;
- paired-hidden-world privacy fixtures.

No prompt answer/model inference in I3.

## I4 — flat prompt families

Owns complete source-order candidate projection + exact private response binding for:

```text
SELECT_EFFECTYN
SELECT_YESNO
SELECT_OPTION
SELECT_POSITION
SELECT_CHAIN
SELECT_IDLECMD
SELECT_BATTLECMD
```

For every supported flat prompt:

```text
N protocol options
= N adapter candidates
= N future model candidate rows
```

Preserve source order and duplicates. No N=1 auto-answer.

## I5 — combinatorial prompt continuations

Owns adapter-local continuation state machines for:

```text
SELECT_CARD
SELECT_TRIBUTE
SELECT_SUM
SELECT_PLACE
SELECT_DISFIELD
SELECT_COUNTER
SORT_CARD
SORT_CHAIN
ANNOUNCE_RACE
ANNOUNCE_ATTRIB
ANNOUNCE_NUMBER
SELECT_UNSELECT_CARD
```

Rules:

- every current continuation domain is complete;
- prior picks may constrain later domains;
- no legal terminal completion consistent with prior choices may be truncated/fabricated/deduplicated/unreachable;
- no intermediate CTOS response;
- exactly one final original-protocol response at terminal continuation completion.

`SELECT_SUM` requires independent exact oracle coverage.

`ANNOUNCE_CARD` remains fail-closed unsupported until separately certified.

## I6 — OCGForge model-contract bundle + byte-exact cross-oracle

**Dependency barrier.**

I6 cannot start compatibility claims until OCGForge has the relevant frozen model-contract bundle / accepted baseline artifacts.

Require byte-exact controlled equality across at least:

```text
public state DTO/bytes
public observation identity
semantic locator table
candidate count/order/fields
public_action_key
candidate-domain identity
LogicalModelInputV1
EncodedModelInputV1
numeric state/candidate rows
model-input identity
```

First unexplained byte mismatch => STOP. Never normalize around it while retaining the old contract identity.

## I7 — model-runner IPC + checkpoint binding

Framed child-process IPC, exact contract/checkpoint handshake, one outstanding request, exact N→N finite scores, deterministic selection/tie rule, stale/late/duplicate/wrong-domain/wrong-checkpoint rejection, timeout/crash fail closed, no retry/fallback.

## I8 — one prompt end-to-end

Exactly one flat prompt family first, from EDOPro packet through public state/candidates/model runner/selected key/final CTOS response.

No scope jump to all prompts in one task.

## I9 — locked-matchup reachable coverage / complete local duels

Both locked decks, both starting-player assignments, reachable supported prompt families, isolated bot-vs-bot adapter/runner processes, deterministic failure on unsupported prompt families, complete local/private duels.

## I10 — audit / replay / first-divergence

Separate:

- public model audit trace;
- perspective-private evaluation trace;
- optional raw protocol trace.

Support deterministic replay and first-divergence diagnosis without making traces gameplay authority or training data by default.

## I11 — Windows release / controlled human readiness

Self-contained Windows application, shared headless/application-controller configuration, isolated model runner, exact manifests/checksums/SBOM/licenses, clean Windows VM acceptance, controlled local/private human-vs-bot readiness.

The UI is configuration/status only; it never owns gameplay legality, candidate decisions, protocol responses, or model authority.

Public servers remain unauthorized unless OI-N01 is separately opened and accepted.

---

# Compatibility vocabulary

Never use one vague `checkpoint compatible` boolean. Track separately:

1. `CHECKPOINT_BINARY_COMPATIBILITY`
2. `INPUT_CONTRACT_COMPATIBILITY`
3. `RULES_DOMAIN_COMPATIBILITY`
4. `EVALUATION_COMPARABILITY`

Until I6 proves otherwise:

```text
CHECKPOINT_BINARY_COMPATIBILITY=UNPROVEN
INPUT_CONTRACT_COMPATIBILITY=UNPROVEN
RULES_DOMAIN_COMPATIBILITY=DIFFERENT_OR_UNPROVEN
EVALUATION_COMPARABILITY=UNPROVEN
```

The current Project-Ignis runtime must never reuse OCGForge's historical rules-bundle identity for different runtime contents.

---

# Acceptance-gate families

Maintain dedicated gates for:

- protocol framing/version/order/failure;
- privacy/hidden-world equality/leakscan;
- complete candidate domains and continuation reachability;
- exact response binding;
- deterministic inference/no fallback;
- complete local gameplay coverage;
- provenance/runtime identities;
- Windows release integrity;
- public-server permission.

Generated acceptance evidence is derived data and must never be hand-edited to make a task pass.

---

# Cross-project sequencing

OCGForge-Ignis I2–I5 may progress in parallel with OCGForge Phase-6 Task5 work.

```text
OCGForge-Ignis:
I2 → I3 → I4 → I5

OCGForge:
Task5 → Task6 → Task7
```

I6 waits for the relevant OCGForge model-contract/baseline freeze. I7+ follows I6 acceptance.

See the core/ML roadmap for Task5 evaluation, PyTorch/JAX bakeoff, first BC baseline, Kaggle dataset scaling, RL, and self-play:

- https://github.com/chrismaghuhn/OCGForge/issues/16

---

# Stable invariants

- fail closed when protocol/legality/privacy/identity/replay behavior is unproven;
- never truncate/fabricate/deduplicate/sort legal domains to keep a duel alive;
- semantic identity must not depend on TCP segmentation, pointer, PID, wall time, task/thread scheduling, filesystem paths, mutable aliases, or unordered iteration;
- private response bindings and protocol locators never become model features;
- public and private traces remain explicitly separated;
- no silent upstream-pin changes;
- no copied EDOPro/WindBot implementation code;
- performance changes require measured semantic equivalence;
- human/external matches are evaluation/debug data unless separately admitted under a future versioned training-data contract.
